BTrDB.cs

```diff
@@ -1,131 +1,732 @@
+            this.currentUpstream.Dispose();
+        }
+
+        /// <summary>
+        /// Implement IEnumerable
+        /// </summary>
+        public IEnumerator<RetT> GetEnumerator()
+        {
+            if (consumed)
+            {
+                throw new InvalidOperationException("a cursor can only be consumed once");
+            }
+            consumed = true;
+            return this;
+        }
+
+        /// <summary>
+        /// Implement IEnumerator
+        /// </summary>
+        public bool MoveNext()
+        {
+            Task<bool> rv = MoveNextAsync();
+            rv.Wait();
+            return rv.Result;
+           /*
+            * TODO remove dead code
+            if (currentUpstream == null || !currentUpstream.MoveNext())
+            {
+                var rv = src.MoveNext(ctx.Token);
+                rv.Wait(ctx.Token);
+                if (rv.Result == false)
+                {
+                    return false;
+                }
+                currentUpstream = converter(src.Current).GetEnumerator();
+                return currentUpstream.MoveNext();
+            }
+            return true;*/
+        }
+
+        /// <summary>
+        /// This is an async enumerator function, which is not currently exposed but could be useful later
+        /// </summary>
+        public async Task<bool> MoveNextAsync()
+        {
+            if (currentUpstream == null || !currentUpstream.MoveNext())
+            {
+                bool rv = await src.MoveNext(ctx.Token).ConfigureAwait(false);
+                if (rv == false)
+                {
+                    return false;
+                }
+                currentUpstream = converter(src.Current).GetEnumerator();
+                return currentUpstream.MoveNext();
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Implement IEnumerator. Reset() is not supported.
+        /// </summary>
+        public void Reset()
+        {
+            throw new InvalidOperationException("BTrDB cursors can not be reset");
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+    }
+
+    
+    /// <summary>
+    /// Stream represents a stream in BTrDB, it is the primary way of inserting and querying data
+    /// </summary>
     public class Stream
     {
-        private Guid id;
-        private BTrDB db;
-        internal Stream(Guid id, BTrDB db)
+        /// <summary>
+        /// Handle to the GRPC conn
+        /// </summary>
+        private readonly BTrDB db;
+
+        /// <summary>
+        /// The UUID of the stream
+        /// </summary>
+        private readonly Guid id;
+       
+        /// <summary>
+        /// The collection. May be null
+        /// </summary>
+        private string collection;
+
+        /// <summary>
+        /// The tags for the stream, may be null
+        /// </summary>
+        private Dictionary<string, string> tags;
```

Comment:
Just for my own reference, does C# allow null strings so we don't have to deal with any of that pointer business?

Reply:
yep